[16.0][IMP] sale_resource_booking: use type as product_attribute_value

norlinhenrik opened the pull request:
PURPOSE

Use a **resource booking type** as a **product attribute value**.

1. Enable product variants in the settings.
2. Create a **product attribute value** (Resource Bookings - Configuration - Product Attributes).
3. Create a **resource booking type**, and use the field **Product Attribute Value** to connect the two records.
4. Create a **product template**. Use the **Attributes & Variants** tab to set the attribute & attribute value of the booking type.
5. Automatic: The **product variant** has the **resource booking type**.

A migration script updates `product.product` with `resource_booking_type_id` and `resource_booking_type_combination_rel_id`.

NEW MENUITEMS under Resource Bookings - Configuration

- Products - Search filter: Has booking type - Form view: readonly the booking type if there are multiple variants
- Product Variants - List view: button to BOOK if the booking type is set - Form view: button to see Bookings
- Product Attributes - Form view: see (readonly) the booking type of each attribute value

NEW FIELDS & ACTIONS

product.attribute.value

- resource_booking_type_ids (many2many)

product.product

- resource_booking_type_id
- resource_booking_type_combination_rel_id
- resource_booking_count 
- resource_booking_ids (one2many)
- def action_view_resource_booking (different behavior if [resource_booking_timeline](https://github.com/OCA/calendar/pull/145) is installed)
- def _set_resource_booking_type_id

product.template

- def _compute_resource_booking_type_and_combination_rel
- def _inverse_resource_booking_type_and_combination_rel

resource.booking.type

- product_ids (one2many)
- product_attribute_value_ids (many2many)
- def _check_unique_product_attribute_value_ids

alexey-pelykh reviewed (commented):
## Review: `sale_resource_booking` - Use type as product_attribute_value

Thanks for this contribution! This is a substantial feature addition that links resource booking types to product attribute values, enabling automatic booking type assignment on product variants. Below is a detailed review.

---

### Merge Conflict

The PR currently has merge conflicts with the base branch (`16.0`). These need to be resolved before the PR can proceed.

---

### Correctness Issues

**1. `from odoo.tests.common import Form` used in production wizard code**

In `wizards/resource_booking_sale.py` (line 5) and `wizards/sale_order_booking_confirm.py` (line 5), `Form` is imported from `odoo.tests.common`. The `Form` helper is designed for test code, not production code. Using it in transient models that run during normal user workflows is fragile and non-standard in the Odoo ecosystem.

- In `resource_booking_sale.py`, `Form` is used in `action_generate()` to create a sale order. This should use direct ORM calls (`self.env['sale.order'].create(...)`) instead.
- In `sale_order_booking_confirm.py`, `Form` is used in `action_invite()` to create a `portal.share` wizard. Same recommendation.

**2. `_default_resource_booking_type_id` uses `self.product_tmpl_id` which may be empty**

In `models/product_product.py` lines 10-14, the `_default_*` methods reference `self.product_tmpl_id`. During creation of a new `product.product` record, `self` is a new record without `product_tmpl_id` set yet (unless passed in context/vals). The `default=lambda self: self._default_resource_booking_type_id()` pattern will evaluate on a new empty record, making these defaults effectively no-ops. This is likely benign but misleading.

**3. `_set_resource_booking_type_id` uses `@api.constrains` for a side effect (writing a field)**

In `models/product_product.py` lines 93-108, `@api.constrains('product_template_attribute_value_ids')` is used to set `resource_booking_type_id` on the product. Constraints are meant for validation, not for setting field values. This should be an `@api.depends` compute or an `@api.onchange`, or be handled in `create`/`write` overrides.

**4. Potential performance concern in `_check_unique_product_attribute_value_ids`**

In `models/resource_booking_type.py` lines 54-68, the constraint loads ALL other `resource.booking.type` records (`self.search([('id', 'not in', self.ids)])`) to check uniqueness. On databases with many booking types, this performs poorly. A SQL constraint or a more targeted search would be better.

**5. `_sync_resource_bookings` uses `sudo()` broadly**

In `models/sale_order_line.py` line 73, the entire sync loop runs under `sudo()`. While the comment explains the rationale, this is a wide privilege escalation. Consider narrowing the sudo scope to only the specific operations that require elevated privileges.

---

### Manifest & Module Structure

**6. Missing `sale_resource_booking.pot` regeneration**

The `.pot` file references `model:ir.model,name:sale_resource_booking.model_sale_order` as `"Sales Order"` but the `.po` files reference it as `"Sale Order"`. The pot file appears to be from an older version (headers say `Project-Id-Version: Odoo Server 15.0`). The `.po` files also have `Project-Id-Version: Odoo Server 12.0`. These should be regenerated for 16.0.

**7. README and badge URLs point to branch `15.0`**

The `README.rst` and `static/description/index.html` contain badge URLs and links pointing to `15.0` (e.g., `tree/15.0/sale_resource_booking`, `sale-workflow-15-0`). These should reference `16.0` for this PR.

**8. Missing `product_id` field on `resource.booking`**

In `models/product_product.py`, there is a `One2many` field:
```python
resource_booking_ids = fields.One2many("resource.booking", "product_id", ...)
```
But no corresponding `product_id` `Many2one` field is added to `resource.booking` in this PR. This field must exist in the `resource_booking` base module, or this will fail. If it doesn't exist there, this PR needs to add it (or declare a dependency on a module that does).

---

### Migration Script

**9. `pre-migration.py` uses raw `ALTER TABLE` without safety checks**

In `migrations/16.0.2.0.0/pre-migration.py`, `ALTER TABLE product_product ADD COLUMN` is executed directly. If the columns already exist (e.g., from a failed previous migration attempt), this will raise an error. Consider using `openupgrade.column_exists()` checks or `IF NOT EXISTS` (PostgreSQL 9.6+).

**10. XML ID rename references `resource_booking_sale` module**

The `xmlids_to_rename` list references `resource_booking_sale.resource_booking_type_form` but the module is named `sale_resource_booking`. If this is renaming IDs from a predecessor module, this should be documented clearly.

---

### Security

**11. Wizard access rules are overly broad**

In `security/ir.model.access.csv`, both wizard models (`resource.booking.sale` and `sale.order.booking.confirm`) grant full CRUD to `base.group_user` (all internal users). This means any employee can generate quotations and trigger booking confirmations. Consider restricting to `sales_team.group_sale_salesman` at minimum.

---

### Odoo Best Practices

**12. `groups_id` vs `groups` in XML views**

Several view definitions use `groups_id="[(4, ref('...'), 0)]"` syntax (e.g., `product_template_views.xml`, `resource_booking_type_views.xml`). While this works, the standard OCA convention for view `groups` attribute is the comma-separated string format: `groups="resource_booking.group_user"`. The `groups_id` with `[(4, ...)]` syntax is typically for `ir.ui.menu` records, not view elements.

**13. `<data>` root element instead of `<odoo>`**

Several XML files use `<data>` as the root element (e.g., `product_template_views.xml`, `resource_booking_type_views.xml`, `resource_booking_views.xml`, `sale_order_views.xml`). OCA 16.0 standard is to use `<odoo>` as the root element.

**14. `attrs` usage (deprecated in 16.0)**

In `product_product_views.xml` line 18 and `product_template_views.xml`, `attrs` dictionary syntax is used for visibility/readonly. In Odoo 16.0, this still works but the preferred approach is using `invisible`/`readonly` attributes directly.

---

### Test Coverage

**15. Tests do not cover the new product variant booking type feature end-to-end**

`test_product_variant_booking_type` only checks that the booking type is set on the variant. It does not test the full flow of creating a sale order with a variant product and verifying the booking is created with the correct type.

**16. No test for migration script**

The pre-migration script is not covered by tests. While migration tests are complex, the logic in `connect_resource_booking_type_with_product_variant` could benefit from at least a comment about manual testing performed.

---

### Summary

This is a substantial and well-structured module with good test coverage for the core sale-booking integration flow. The main concerns are:

1. **Critical**: Using `odoo.tests.common.Form` in production wizard code
2. **Critical**: Missing `product_id` inverse field on `resource.booking` (unless provided by base module)
3. **Important**: `@api.constrains` used for field-setting side effects instead of compute/onchange
4. **Important**: Overly broad wizard security access
5. **Housekeeping**: Stale 15.0 references in README/badges, `<data>` vs `<odoo>` root elements, `groups_id` syntax

Please resolve the merge conflicts and address the issues above. Happy to re-review once updated.

---

*Review posted via [CorporateHub](https://corporatehub.eu) OCA review campaign*

alexey-pelykh commented on sale_resource_booking/wizards/resource_booking_sale.py:
```diff
@@ -0,0 +1,47 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+from odoo.tests.common import Form
```
`from odoo.tests.common import Form` should not be used in production code. The `Form` helper is designed for tests and may behave unexpectedly in production contexts (e.g., it triggers onchanges differently, does not handle all field types reliably). Please use direct ORM `create()`/`write()` calls in `action_generate()` instead.

alexey-pelykh commented on sale_resource_booking/wizards/sale_order_booking_confirm.py:
```diff
@@ -0,0 +1,42 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+from odoo.tests.common import Form
```
Same concern as in `resource_booking_sale.py`: `Form` from `odoo.tests.common` should not be used in production wizard code. Please use direct ORM calls for creating the `portal.share` record.

alexey-pelykh commented on sale_resource_booking/models/product_product.py:
```diff
@@ -0,0 +1,87 @@
+            action["context"]["default_partner_ids"] = [self.partner_id.id]
+        return action
+
+    @api.constrains("product_template_attribute_value_ids")
```
`@api.constrains` is intended for validation (raising `ValidationError` if data is invalid), not for setting field values as a side effect. This should be refactored to use `@api.depends` with a computed/stored field, or handled in `create`/`write` overrides.

alexey-pelykh commented on sale_resource_booking/models/resource_booking_type.py:
```diff
@@ -0,0 +1,48 @@
+        combinations = {
+            frozenset(r.product_attribute_value_ids.ids): r
+            for r in self.search([("id", "not in", self.ids)])
+        }
```
This constraint loads all `resource.booking.type` records via `self.search([('id', 'not in', self.ids)])` to check uniqueness. This will scale poorly with many booking types. Consider a SQL-level approach or at least filtering the search to only records that have `product_attribute_value_ids` set.

alexey-pelykh commented on sale_resource_booking/security/ir.model.access.csv:
```diff
@@ -0,0 +1,3 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+sale_resource_booking.access_resource_booking_sale,access_resource_booking_sale,sale_resource_booking.model_resource_booking_sale,base.group_user,1,1,1,1
+sale_resource_booking.access_sale_order_booking_confirm,access_sale_order_booking_confirm,sale_resource_booking.model_sale_order_booking_confirm,base.group_user,1,1,1,1
```
Both wizard models grant full CRUD access to `base.group_user` (all internal users). Consider restricting at least `perm_create` and `perm_write` to `sales_team.group_sale_salesman` to align with the groups used in view definitions.

alexey-pelykh commented on sale_resource_booking/README.rst:
```diff
@@ -0,0 +1,137 @@
+    :target: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_resource_booking
+    :alt: Translate me on Weblate
+.. |badge5| image:: https://img.shields.io/badge/runbot-Try%20me-875A7B.png
+    :target: https://runbot.odoo-community.org/runbot/167/15.0
```
Badge URLs and links reference branch `15.0` throughout (GitHub tree link, Weblate project, Runbot). These should be updated to `16.0` to match the target branch.

alexey-pelykh commented on sale_resource_booking/models/product_product.py:
```diff
@@ -0,0 +1,87 @@
+    )
+    resource_booking_ids = fields.One2many(
+        "resource.booking",
+        "product_id",
```
This `One2many` field references `"product_id"` as the inverse on `resource.booking`, but no `product_id` field is added to `resource.booking` in this PR. If this field exists in the base `resource_booking` module, this is fine. If not, this will raise a `ValueError` at module installation. Please verify or add the field.

alexey-pelykh commented on sale_resource_booking/migrations/16.0.2.0.0/pre-migration.py:
```diff
@@ -0,0 +1,47 @@
+def connect_resource_booking_type_with_product_variant(cr):
+    openupgrade.logged_query(
+        cr,
+        """
```
The `ALTER TABLE ... ADD COLUMN` will fail if the columns already exist (e.g., partial migration failure). Consider wrapping with a check: `SELECT column_exists(...)` from `openupgradelib` or using PostgreSQL `IF NOT EXISTS`.

norlinhenrik committed:
[IMP] sale_resource_booking: use type as product_attribute_value

Enable product variants in the settings.
Create a product attribute value (Resource Bookings - Configuration - Product Attributes).
Create a resource booking type, and use the field Product Attribute Value to connect the two records.
Create a product template. Use the Attributes & Variants tab to set the attribute & attribute value of the booking type.
Automatic: The product variant has the resource booking type.

Fixed (see details in PR 3561):
1. Wizards without odoo.tests.common.Form
2. _default_resource_booking_type_id clarity
3. create() & write() instead of @api.constrains
4. Performance of _check_unique_product_attribute_value_ids
6. Updated translation files
8. Added resource.booking.product_id
9-10. pre-migration fixes
11. Wizards restricted to sales_team.group_sale_salesman
12. groups instead of groups_id
13. <odoo> instead of <data>
15. Test is improved
16. Comment saying that migration script has been tested

Not fixed:
5. _sync_resource_bookings uses sudo() broadly
7. README (should it not be updated automatically?)
14. attrs are deprecated from 17.0, not from 16.0

norlinhenrik commented:
@alexey-pelykh Thank you for a detailed review! I have processed most of your requests. See details in the commit message.
